main.cpp

```diff
@@ -90,199 +83,268 @@ bool generateKeyPair(ethsnarks::ProtoboardT& pb, std::string& baseFilename)
+
+    struct ProverStatus
+    {
+        bool proving = false;
+        std::string blockFilename;
+        std::string proofFilename;
+    };
+
+    struct ProverStatusRAII
     {
-        json jWithdrawals = input["withdrawals"];
-        if (jWithdrawals.size() != numWithdrawals)
+        ProverStatus& proverStatus;
+        ProverStatusRAII(ProverStatus& _proverStatus, const std::string& blockFilename, const std::string& proofFilename) : proverStatus(_proverStatus)
         {
-            std::cerr << "Invalid number of withdrawals in input file: " << jWithdrawals.size() << std::endl;
-            return false;
+            proverStatus.proving = true;
+            proverStatus.blockFilename = blockFilename;
+            proverStatus.proofFilename = proofFilename;
         }
 
-        Loopring::OffchainWithdrawalBlock block = input.get<Loopring::OffchainWithdrawalBlock>();
-
-        // Generate witness values for the given input values
-        if (!circuit.generateWitness(block))
+        ~ProverStatusRAII()
         {
-            std::cerr << "Could not generate witness!" << std::endl;
-            return false;
+            proverStatus.proving = false;
         }
-    }
-    return true;
-}
+    };
 
-bool cancel(Mode mode, bool onchainDataAvailability, unsigned int numCancels, const json& input, ethsnarks::ProtoboardT& outPb)
-{
-    // Build the circuit
-    Loopring::OrderCancellationCircuit circuit(outPb, "circuit");
-    circuit.generate_r1cs_constraints(onchainDataAvailability, numCancels);
-    circuit.printInfo();
+    // Load the proving key a single time
+    auto proving_key = loadProvingKey(provingKeyFilename);
 
-    if (mode == Mode::Validate || mode == Mode::Prove)
-    {
-        json jCancels = input["cancels"];
-        if (jCancels.size() != numCancels)
+    // Prover status info
+    ProverStatus proverStatus;
+    // Lock for the prover
+    std::mutex mtx;
+    // Setup the server
+    Server svr;
+    // Called to prove blocks
+    svr.Get("/prove", [&](const Request& req, Response& res) {
+        const std::lock_guard<std::mutex> lock(mtx);
+
+        // Parse the parameters
+        std::string blockFilename = req.get_param_value("block_filename");
+        std::string proofFilename = req.get_param_value("proof_filename");
+        std::string strValidate = req.get_param_value("validate");
+        bool validate = (strValidate.compare("true") == 0) ? true : false;
+        if (blockFilename.length() == 0)
         {
-            std::cerr << "Invalid number of cancels in input file: " << jCancels.size() << std::endl;
-            return false;
+            res.set_content("Error: block_filename missing!\n", "text/plain");
+            return;
         }
 
-        Loopring::OrderCancellationBlock block = input.get<Loopring::OrderCancellationBlock>();
+        // Set the prover status for this session
+        ProverStatusRAII statusRAII(proverStatus, blockFilename, proofFilename);
```

Comment:
Easy state management. There's multiple return call where we'd need to set `proving = false`, but by using an object that automatically gets out of scope when the function ends we can just do it once in the destructor. Basically it makes it much less error prone because it's hard to keep track otherwise where the state needs to be modified. `std::lock_guard` for the mutex actually works in the same way to make sure the mutex is unlocked when the function ends.

- https://en.cppreference.com/w/cpp/language/raii